### ISSUE TITLE
Remote API

### ISSUE THREAD

**Issue by jjh42** [facebook/nuclide#516]
Nuclide Remote is fantastic. However, much like Atom, Nuclide will never be all things to all people.

The solution, obviously, is to allow people to develop packages to solve their particular itch. I've modified a few packages to run commands over ssh to make them work with Nuclide, but its always a bit hacky.

It would be great if Nuclide-Remote would explicitly expose API for other packages to use to make it easy for packages to be compatible with remote files. e.g. remoteExec (run a linter, compiler etc. on the remote associated with this package, ideally this would also just run locally in the case where a textbuffer is locally), remotePath (currently I hackily import nuclide-uri) would be a great start.

**Comment by nmote** [comment]
Yes, this is definitely something we want to do. It's been inching higher on our priority list but we still have no concrete plans to make it happen. @peterhal wrote up [an initial spec](https://github.com/facebook/nuclide/blob/master/pkg/nuclide-rpc/docs/InstallableRemoteModules.md) for this.

Specifically, though, we will probably require users to manually install the npm module on the remote server for two reasons: (1) we want to avoid avenues for remote code execution, and (2) a lot of things can go wrong with `npm install` that we don't want to have to account for in our UI.

**Comment by jjh42** [comment]
Can't any atom package already access the ssh connection during the initial connection and execute commands?

Anyway, at this stage most of the packages I've messed with are doing everything on the client other than executing a command (e.g. a linter) so they haven't required installing a package on the server, but I suppose that's not a particularly onerous requirement if it fits the security model better, it would be nice to have an example of a remote package that simply executes and command and returns the respond.

**Comment by nmote** [comment]
Yes, but we don't want to make it any worse than it already is 😄 And we certainly don't want to build something that is inherently insecure by design when there is a reasonable alternative.

For the same reason, we won't be bundling something that allows arbitrary remote commands to be executed, like you're suggesting, but if somebody used this interface to create a third party package that did that, we couldn't stop them.

**Comment by capouch** [comment]
Would the thrust of this thread indicate that at least as things stand I cannot expect to be able to do git commands on a remote-located project?  I have git-plus installed and it doesn't seem to be enabled when I open remotely.

**Comment by jjh42** [comment]
@capouch in general any package that is specially made to do so (which is not terribly easy right now) and isn't one of the built-in nuclide packages isn't going to work with remote projects.

**Comment by alanzanattadev** [comment]
@capouch @jjh42 you can use my git packages https://atom.io/users/alanzanattadev

**Comment by nmote** [comment]
As long as you understand that this should not be considered a stable API...

**Comment by viveksjain** [comment]
+1. When this is implemented, I think it would be very helpful if nuclide comes with a `fs` module that is a drop-in replacement for the system one, but also works with `nuclide:` paths.
